src/openrct2/management/NewsItem.cpp

```diff
@@ -207,73 +207,72 @@ static int32_t news_item_get_new_history_slot()
+CoordsXYZ news_item_get_subject_location(int32_t type, int32_t subject)
 {
-    Ride* ride;
-    Peep* peep;
-    Vehicle* vehicle;
+    CoordsXYZ subjectLoc;
 
     switch (type)
     {
         case NEWS_ITEM_RIDE:
-            ride = get_ride(subject);
+        {
+            Ride* ride = get_ride(subject);
             if (ride == nullptr || ride->overall_view.isNull())
             {
-                *x = LOCATION_NULL;
+                subjectLoc.setNull();
                 break;
             }
-            *x = ride->overall_view.x + 16;
-            *y = ride->overall_view.y + 16;
-            *z = tile_element_height({ *x, *y });
+            auto rideViewCentre = ride->overall_view.ToTileCentre();
+            subjectLoc = CoordsXYZ{ rideViewCentre, tile_element_height(rideViewCentre) };
             break;
+        }
         case NEWS_ITEM_PEEP_ON_RIDE:
-            peep = GET_PEEP(subject);
-            *x = peep->x;
-            *y = peep->y;
-            *z = peep->z;
-            if (*x != LOCATION_NULL)
+        {
+            Peep* peep = GET_PEEP(subject);
+            subjectLoc = CoordsXYZ{ peep->x, peep->y, peep->z };
+            if (!subjectLoc.isNull())
                 break;
 
             if (peep->state != 3 && peep->state != 7)
             {
-                *x = LOCATION_NULL;
+                subjectLoc.setNull();
                 break;
             }
 
             // Find which ride peep is on
-            ride = get_ride(peep->current_ride);
+            Ride* ride = get_ride(peep->current_ride);
             if (ride == nullptr || !(ride->lifecycle_flags & RIDE_LIFECYCLE_ON_TRACK))
             {
-                *x = LOCATION_NULL;
+                subjectLoc.setNull();
                 break;
             }
 
             // Find the first car of the train peep is on
-            vehicle = &(get_sprite(ride->vehicles[peep->current_train])->vehicle);
+            Vehicle* vehicle = &(get_sprite(ride->vehicles[peep->current_train])->vehicle);
             // Find the actual car peep is on
             for (int32_t i = 0; i < peep->current_car; i++)
             {
                 vehicle = &(get_sprite(vehicle->next_vehicle_on_train)->vehicle);
             }
-            *x = vehicle->x;
-            *y = vehicle->y;
-            *z = vehicle->z;
+            subjectLoc = CoordsXYZ{ vehicle->x, vehicle->y, vehicle->z };
             break;
+        }
         case NEWS_ITEM_PEEP:
-            peep = GET_PEEP(subject);
-            *x = peep->x;
-            *y = peep->y;
-            *z = peep->z;
+        {
+            Peep* peep = GET_PEEP(subject);
+            subjectLoc = CoordsXYZ{ peep->x, peep->y, peep->z };
             break;
+        }
         case NEWS_ITEM_BLANK:
-            *x = subject;
-            *y = subject >> 16;
-            *z = tile_element_height({ *x, *y });
+        {
+            auto subjectXY = CoordsXY{ subject, subject >> 16 };
```

Comment:
This look wrong. I think someone has broken this in the past. I'm going to guess subject in this case has been created from packing to int16_t's into a int32_t. Eitherway this new code and the old code is wrong because x needs to be &'ed with 0xFFFF.

Reply:
And what to do with the `y`?

Reply:
No need to change the y, as you already did a bitshift by 16, which ensures that the upper 16 bits are 0 anyway.